File: Year 2/Semester 2/MPP/JavaSocket/AirlineServer/src/main/java/airline/server/impl/AirlineServicesImpl.java
package airline.server.impl;

import airline.model.Employee;
import airline.model.Flight;
import airline.model.Ticket;
import airline.persistence.IEmployeeRepository;
import airline.persistence.IFlightRepository;
import airline.persistence.ITicketRepository;
import airline.services.IAirlineObserver;
import airline.services.IService;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

public class AirlineServicesImpl implements IService {
    private IEmployeeRepository employeeRepository;
    private ITicketRepository ticketRepository;
    private IFlightRepository flightRepository;

    private Map<Integer, IAirlineObserver> loggedEmployees;


    public AirlineServicesImpl(IEmployeeRepository employeeRepository, ITicketRepository ticketRepository, IFlightRepository flightRepository) {
        this.employeeRepository = employeeRepository;
        this.ticketRepository = ticketRepository;
        this.flightRepository = flightRepository;
        loggedEmployees =new ConcurrentHashMap<>();
    }

    public synchronized Employee login(Employee employee, IAirlineObserver client) throws Exception {
        Employee employeeToLogin=employeeRepository.getEmployee(employee.getUsername(),employee.getPassword());
        if (employeeToLogin!=null){
            if (loggedEmployees.get(employeeToLogin.getId())!=null)
                throw new Exception("Employee already logged in.");

            loggedEmployees.put(employeeToLogin.getId(),client);

        }
        else
            throw new Exception("Authentication failed.");
        return employeeToLogin;
    }

    public synchronized void logout(Employee employee, IAirlineObserver client) throws Exception {
        IAirlineObserver loggedEmployee= loggedEmployees.remove(employee.getId());
        if (loggedEmployee==null)
            throw new Exception("Employee "+String.valueOf(employee.getId())+" is not logged in.");
    }
    @Override
    public synchronized Iterable<Flight> getAllFlights() {
        return flightRepository.getAll();
    }

    public synchronized int getNumberOfTicketsForFlight(Integer flightId) throws Exception{
        List<Ticket> tickets = (List<Ticket>) ticketRepository.getAll();
        List<Ticket> filtered_tickets = tickets.stream()
                .filter(ticket -> ticket.getFlightId() == flightId)
                .collect(Collectors.toList());

        return filtered_tickets.size();
    }

    public synchronized Iterable<Flight> getFlights(String destination, String date) throws Exception{
        List<Flight> flights = (List<Flight>) flightRepository.getAll();
        List<Flight> filtered_flights = null;
        if (date.equals(""))
            filtered_flights = flights.stream()
                    .filter(flight -> {
                        try {
                            return flight.getDestination().equals(destination) && flight.getAvailableSeats() > 0;
                        } catch (Exception e) {
                            throw new RuntimeException(e);
                        }
                    })
                    .collect(Collectors.toList());
        else {
            filtered_flights = flights.stream()
                    .filter(flight -> {
                        try {
                            return flight.getDestination().equals(destination) && flight.getDate().equals(date) &&  flight.getAvailableSeats() > 0;
                        } catch (Exception e) {
                            throw new RuntimeException(e);
                        }
                    })
                    .collect(Collectors.toList());
        }
        return filtered_flights;
    }


    public synchronized Employee getEmployee(String username, String password) {
        Employee employee = employeeRepository.getEmployee(username, password);
        return employee;
    }

    public synchronized void buyTicket(int flightId, String tourist_name, String client_address) {
        Ticket ticket = new Ticket(tourist_name, client_address, flightId);
        ticketRepository.add(ticket);
        flightRepository.updateNumberOfSeats(flightId);
        for (IAirlineObserver loggedUser: loggedEmployees.values()) {
            loggedUser.ticketBought(ticket);
        }
    }
}

File: Year 2/Semester 2/MPP/Protobuf/ProtoCSharp/AirlineServer/ServerAirline/AirlineServicesImpl.cs
using Model;
using Persistency;
using Services;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Server
{
    public class AirlineServicesImpl : IService
    {
        private readonly IEmployeeRepository employeeRepository;
        private readonly ITicketRepository ticketRepository;
        private readonly IFlightRepository flightRepository;

        private readonly IDictionary<int, IObserver> loggedEmployees;



        public AirlineServicesImpl(IEmployeeRepository employeeRepository, ITicketRepository ticketRepository, IFlightRepository flightRepository)
        {
            this.employeeRepository = employeeRepository;
            this.ticketRepository = ticketRepository;
            this.flightRepository = flightRepository;

            loggedEmployees=new Dictionary<int, IObserver>();
        }


        public void buyTicket(int flightId, string tourist_name, string client_address)
        {
            Ticket ticket = new Ticket(tourist_name, client_address, flightId);
            ticketRepository.add(ticket);
            flightRepository.updateNumberOfSeats(flightId);
            foreach (var loggedEmployee in loggedEmployees.Values)
            {
                Task.Run(() => loggedEmployee.TicketBought(ticket));
            }
            
            
            

        }

        public List<Flight> getAllFlights()
        {
            List<Flight> flights = flightRepository.GetAll().Where(f => f.NumberOfSeats - getNumberOfTicketsForFlights(f.Id) > 0).ToList();
            return flights;

        }

        public Employee getEmployee(string username, string password)
        {
            List<Employee> employees = employeeRepository.GetAll().ToList();
            Employee employee = employees.Where(e => e.Username == username && e.Password == password).FirstOrDefault();
            if (employee != default)
            {
                return employee;
            }
            else return null;

        }

        public List<Flight> getFlights(string destination, string date)
        {
            List<Flight> flights = flightRepository.GetAll().ToList();
            return flights.Where(f => f.Destination.StartsWith(destination) && f.Date == date && f.NumberOfSeats - getNumberOfTicketsForFlights(f.Id) > 0).ToList();

        }

        public int getNumberOfTicketsForFlights(int flightId)
        {
            List<Ticket> tickets = ticketRepository.GetAll().ToList();
            return tickets.Where(t => t.FlightId == flightId).Count();
        }

        public Employee login(Employee employee, IObserver client)
        {
            List<Employee> employees = employeeRepository.GetAll().ToList();
            Employee employee1 = employees.Where(e => e.Username == employee.Username && e.Password == employee.Password).FirstOrDefault();

            if (employee1 != null)
            {
                if (loggedEmployees.ContainsKey(employee1.Id))
                    throw new Exception("Already logged in");
                loggedEmployees[employee1.Id] = client;
            }
            else throw new Exception("Failed login");
            return employee1;
        }

        public void logout(Employee employee)
        {
            var Client = loggedEmployees[employee.Id];
            if (Client == null)
                throw new Exception("User " + employee.Id +" is not logged in!");
            loggedEmployees.Remove(employee.Id);
        }

        
    }
}
